[.NET] Better `.editorconfig` setup in `modules/mono/`

paulloz opened the pull request:
- Remove explicit warnings on CA1716, CA1304 (already included in the level of analysis we run)
- Do not silence CA1805
- Properly fix REFL045
- Properly fix CS1572, CS1573
- Suppress CS0618 in `InvokeGodotClassMethod`
- Only silence CA1707, CA1711, CA1720 on a more limited scope
- Temporarily silence CS1734 in generated code
- Enforce (not really) naming conventions
- Enforce (not really) Allman braces

paulloz committed:
Cleanup C# projects, code quality & style

New rules:
- Do not silence CA1805 any more
- Limit where we silence CA1707, CA1711, CA1720
- Enforce severity=warning for IDE0040
- Enforce Allman style braces
- Enforce naming conventions (IDE1006 is still severity=suggestion)

Fixes:
- Fix REFL045, CS1572, CS1573
- Suppress CS0618 when generating `InvokeGodotClassMethod`
- Fix indent when generating GD_constants.cs
- Temporarily silence CS1734 in generated code
- Fix a lot of naming rule violations

Misc.:
- Remove ReSharper comments for RedundantNameQualifier
- Remove suppression attributes for RedundantNameQualifier
- Remove severity=warnings for CA1716, CA1304 (already included in the level of analysis we run)

raulsntos commented on modules/mono/.editorconfig:
```diff
@@ -13,31 +13,33 @@ trim_trailing_whitespace = true
+# Diagnostics to prevent defensive copies of `in` struct parameters
+resharper_possibly_impure_method_call_on_readonly_variable_highlighting = error
+
+# Severity levels for dotnet_naming_rule only affect IDE environments.
```
Do the naming rules affect `dotnet format`? If so, our CI check should enforce them.

paulloz replied:
It does, and doesn't. It's dependent on the severity level we set for IDE1006. We can set it on suggestion if we don't want to block PRs, or on warning if we want to. One thing to note is IDE1006 won't apply changes automatically, so it would be listed as `dotnet format` issues, but no corrections would be visible in the diff. If that make sense?

raulsntos replied:
> it would be listed as `dotnet format` issues, but no corrections would be visible in the diff

But would it show up in the _Files changed_ tab? (like the analyzer tracking diagnostic) If so, I think that may be good enough. But I guess _suggestion_ won't show, so we'd have to go with _warning_.

paulloz replied:
I'm not entirely against setting that as *warning*, but it means we'd need to scrap the current codebase of formatting issues. Also, am I right in the rules I defined? Because it's not entirely consistent at the moment.

raulsntos replied:
At a first glance it looks correct, although they're probably not exhaustive enough (for example it doesn't cover parameters, local variables, properties, etc). For reference, here's the `.editorconfig` used by roslyn: https://github.com/dotnet/roslyn/blob/6eb91fbee2f416285dcc7603e8c8769f78ed0aa8/.editorconfig#L65-L143

As for the current codebase, it's likely that we're not consistent because we didn't start closely following the code-style until somewhat recently. We'll have to do that at some point but I think it's fine to set the severity to _suggestion_ for now.

paulloz replied:
Yes, I only added rules for things where I saw naming issues for now. I'll add more exhaustive ones before un-drafting 👍

raulsntos reviewed (commented):
Looks great, thanks for working on this.

paulloz commented:
I'm un-drafting this one. 

There is an additional commit that I left separate. It fixes the indentation in the generated `GD_constants.cs` file, I needed it to be able to run `dotnet format` without being flooded by bad indentation errors. If I'm not too stupid, content shouldn't have changed, only it doesn't trip the formatter any more. It can either be squashed or dropped.

I wrote additional naming rules in the `.editorconfig` file, but left the diagnostic itself on `suggestion`. I'm linking some logs of a `dotnet format` run with its severity set on `warning` for reference. It's not *that* bad after all.

- [dotnet_format.log](https://github.com/godotengine/godot/files/14366280/dotnet_format.log)
- [dotnet_format.generated.log](https://github.com/godotengine/godot/files/14366283/dotnet_format.generated.log)

raulsntos commented:
From what I can see, most of the violations are from `InteropStructs.cs`, `NativeFuncs.cs` and `Internal.cs`. I think we should just exclude those files.

As for the generated files, it looks like most violations are due to virtual methods having a `_` prefix. I wonder if we can allow that prefix (without it being required) when the method is virtual.

I'm not sure what the rule should be for static readonly, it seems we use `_camelCase` for them a lot but the rules in this PR say it should be `PascalCase`. For reference, the Roslyn repo uses `s_camelCase`. I don't have a strong preference as long as we're consistent, but using `PascalCase` may conflict with other members (for example `Vector3._zero` and `Vector3.Zero`).

It would be great to also add a rule to require tuples to use explicit naming and use PascalCase, there's `dotnet_style_explicit_tuple_names` but it seems we can't specify naming rules for tuple elements yet (https://github.com/dotnet/roslyn/issues/38705).

paulloz commented:
> I'm not sure what the rule should be for static readonly, it seems we use _camelCase for them a lot but the rules in this PR say it should be PascalCase.

Yeah, we kind of use both currently. With the `PascalCase` version more in the source generator part of the codebase. I really don't have a strong opinion, my usual take is to treat them as consts.

paulloz commented:
> As for the generated files, it looks like most violations are due to virtual methods having a _ prefix. I wonder if we can allow that prefix (without it being required) when the method is virtual

Sadly, I don't think that's a thing.

raulsntos commented:
> > As for the generated files, it looks like most violations are due to virtual methods having a `_` prefix. I wonder if we can allow that prefix (without it being required) when the method is virtual
>
> Sadly, I don't think that's a thing.

Looks like there's an issue for this: https://github.com/dotnet/roslyn/issues/46205. Until then, since we should only be using the `_` prefix in generated virtual methods, maybe we can just add an exception for files in the _Generated_ directory.

paulloz commented:
- Removed the rule on `static readonly` (I noticed that currently, private ones are `_camelCase`, and public ones are `PascalCase`, matching the rules for other fields).
- Silence **IDE1006** in the *Generated* directories.
- Ignore **IDE1006** in
  - *NativeFuncs.cs*
  - *NativeFuncs.extended.cs*
  - *InteropStructs.cs*
  - *Internal.cs*
- Did a bit of cleaning (not a lot)
  - Removed a bunch of `[SuppressMessage("ReSharper", "...")]` attribute to replace them with `// ReSharper` comments (I'm not sure if we build releases using `CODE_ANALYSIS` or not)
  - Fixed naming rule violations in *GodotSharp*
  - Fixed naming rule violations in *Godot.SourceGenerators.**

raulsntos commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -1069,29 +1070,27 @@ void BindingsGenerator::_generate_global_constants(StringBuilder &p_output) {
+					 << iconstant.proxy_name
+					 << " = "
+					 << itos(iconstant.value)
+					 << (&iconstant != &last ? ",\n" : "\n");
```
We can just always add a trailing comma, then it looks like we won't need the `last` variable.

```suggestion
					 << ",\n";
```

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs:
```diff
@@ -1253,16 +1258,16 @@ public void Dispose()
 
                 // There's no recursion here, so it's ok to go with a big enough number for most cases
                 // stackMaxSize = stackMaxLength * sizeof(godotsharp_property_def_val_pair)
```
I guess the comment should be updated as well.

```suggestion
                // StackMaxSize = StackMaxLength * sizeof(godotsharp_property_def_val_pair)
```

paulloz commented on modules/mono/glue/GodotSharp/GodotSharp/Core/DelegateUtils.cs:
```diff
@@ -771,10 +771,10 @@ public static godot_variant ConvertToVariant(object? obj)
                 };
 
-            [SuppressMessage("ReSharper", "RedundantNameQualifier")]
+            // ReSharper disable RedundantNameQualifier
```
`disable once` would only affect the next line, here we affect the entire method.

paulloz replied:
Yes, it's because we're inside the `Godot` namespace, and we specify `Godot.Whatever`.

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs:
```diff
@@ -1064,16 +1066,16 @@ public void Dispose()
 
                 // There's no recursion here, so it's ok to go with a big enough number for most cases
                 // stackMaxSize = stackMaxLength * sizeof(godotsharp_property_info)
```
I missed this one before.

```suggestion
                // StackMaxSize = StackMaxLength * sizeof(godotsharp_property_info)
```

raulsntos commented on modules/mono/build_scripts/build_assemblies.py:
```diff
@@ -320,6 +320,7 @@ def generate_sdk_package_versions():
 
     constants = """namespace Godot.SourceGenerators
 {{
+#pragma warning disable IDE0040 // Add accessibility modifiers. This is currently disabled because of https://github.com/dotnet/roslyn/issues/52904
```
We should add a `TODO:` so we can find it easily later, preferably at the start of a comment.

```suggestion
#pragma warning disable IDE0040 // Add accessibility modifiers.
    // TODO: This is currently disabled because of https://github.com/dotnet/roslyn/issues/52904
```

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharpEditor/Compat.cs:
```diff
@@ -7,6 +7,8 @@
 
 namespace Godot;
 
+#pragma warning disable IDE0040 // Add accessibility modifiers. This is currently disabled because of https://github.com/dotnet/roslyn/issues/52904
```
```suggestion
#pragma warning disable IDE0040 // Add accessibility modifiers.
    // TODO: This is currently disabled because of https://github.com/dotnet/roslyn/issues/52904
```

raulsntos commented on modules/mono/glue/GodotSharp/GodotSharp/Compat.cs:
```diff
@@ -8,6 +8,7 @@
 namespace Godot;
 
 #pragma warning disable CS1734 // XML comment on 'X' has a paramref tag for 'Y', but there is no parameter by that name.
+#pragma warning disable IDE0040 // Add accessibility modifiers. This is currently disabled because of https://github.com/dotnet/roslyn/issues/52904
```
```suggestion
#pragma warning disable IDE0040 // Add accessibility modifiers.
    // TODO: This is currently disabled because of https://github.com/dotnet/roslyn/issues/52904
```

raulsntos commented on modules/mono/build_scripts/build_assemblies.py:
```diff
@@ -338,7 +339,7 @@ def generate_sdk_package_versions():
     os.makedirs(generators_dir, exist_ok=True)
 
-    with open(os.path.join(generators_dir, "Common.Constants.cs"), "w") as f:
+    with open(os.path.join(generators_dir, "Common.Constants.cs"), "w", newline="\n") as f:
```
Nice catch. Looking at other python scripts in the codebase, it looks like none of them are specifying the newline character. Also, they tend to specify the encoding (`utf-8`) which we may also want to do here.

paulloz replied:
I'm adding the encoding to both `open()` call we make in this file, I think we care about the newline only when we write sources, to not flood output from `dotnet format` when running on Windows.

raulsntos replied:
From what I see, the other usages of `open()` in the codebase seem to be adding `\n` manually so I think we're good.

paulloz commented:
I'll squash once we know there's nothing more we wanna add to this specific PR.

raulsntos reviewed (approved):
I think this looks good enough already, we can always add more rules to `.editorconfig` later. Thanks for the great cleanup!

paulloz commented:
Here's to hoping I didn't break everything 🥂

akien-mga commented:
Thanks!

aaronfranke commented:
> Enforce (not really) naming conventions

When did the naming convention for private members change to PascalCase?

raulsntos commented:
The changes in this PR are only meant to add the conventions that we've already been following to the `.editorconfig`, so they can be enforced by `dotnet format` in CI. Private fields are still following the `_camelCase` naming that has always been our codestyle.

I guess you're referring to the few constant names that were changed from `_camelCase` to `PascalCase` in this PR. As discussed in the PR, we haven't been consistent throughout the codebase on the constant naming style so we just chose the one that is also recommended by the .NET naming guidelines.
